content/r-package-dev/debugging.Rmd

```diff
@@ -0,0 +1,93 @@
+
+Now we know where the actual error happened, but what if we want to see what led up to it?  Click on the text `getWebServiceData.R#36` in the traceback output to open up the file that contains `getWebServiceData`.  We can set a *breakpoint* in `getWebServiceData` by clicking to the left of the line number where we want it. 
+
+![](../static/img/breakpoint.png) 
+
+Breakpoints pause the function before the line executes, allowing you to inspect environment variables, experiment with other code, etc.  Make the same `readNWISsites` call to see it work.  
+```{r eval = FALSE}
+readNWISsites(siteNumbers = notASite)
+```
+![](../static/img/breakpoint_environment.png)
+
+Notice that the drop down menu in the environment pane now indicates we are in the function environment.  We can now view any objects that currently exist here.  You can run code in the console to test things.  Note if you modify an object here, it could affect the function if you continue to run it.  You have several options to continue executing the function, accessible by the buttons above the console, or their corresponding keyboard shortcuts `n`, `s`, `f`, `c` and `Q`: 
+
+![](../static/img/debugButtons.png)
+
+They will respectively execute the next line of code, step into a function call (going a level deeper), execute the remainder of the function or loop, continue executing everything until a breakpoint is encountered, or exit debugging.  You can see these descriptions by hovering over the buttons.
+
+Click the red breakpoint dot to make it go away.
+
+Note that when you are working on a package from source [better wording here?], you will often need to rebuild it to make breakpoints take effect.  A somewhat more explicit (and non-RStudio-dependent) way to accomplish the same thing is to insert a `browser()` call inside your function.  However, it is important to remember to delete `browser()` statements when you are done with them, so they don't find their way onto your repository!
+
+What if we have a logical error, and therefore don't have a particular line to key in on from `traceback()`?  We can use the `debug()` or `debugonce()`.  These functions insert a `browser` call behind the scenes at the start of a function.  Then you can step through the entire function by individual blocks or lines, using the debugging buttons on the console or the corresponding key commands as shown above.  If you use `debug` instead of `debugonce`, you will need to use `undebug` when you are done to turn of debugging mode. If you click the "Rerun with Debug" button on the error window it will use debug on the function where there error occured.  However, this frequently occurs deep in a base R function, not your code where the actual bug is, and can therefore be less useful than `traceback` or using `debug` yourself. 
```

Comment:
this would be a great place for an example to clarify what a logical error might be and how debugging might help.